File: 卡码/代码随想录/C++/字符串/06-重复的子字符串/LeetCode_0459.cpp
//459. 重复的子字符串
#include <string>
using namespace std;

class Solution {
public:
  bool repeatedSubstringPattern(string s) {
    int n=s.size();
    //法一
//    string s1=s+s;
//    string s2=s1.substr(1,2*n-2);
//    return s2.find(s)!=string::npos;
    //法二
    int next[n];
    next[0]=0;
    int j=0;
    for(int i=1;i<n;i++){
      while(j>0&&s[i]!=s[j]) j=next[j-1];
      if(s[i]==s[j]) j++;
      next[i]=j;
    }
    return next[n-1]!=0&&n%(n-next[n-1])==0;
  }
};

File: 卡码/代码随想录/Python/字符串/string/06-重复的子字符串/LeetCode_0459.py
#459. 重复的子字符串
class Solution:
  def repeatedSubstringPattern(self, s: str) -> bool:
    #法一
    n=len(s)
    # s1=s+s
    # return s1[1:2*n-1].find(s)!=-1
    #法二
    next=[0 for x in range(n)]
    j=0
    for i in range(1,n):
      while j>0 and s[i]!=s[j]:
        j=next[j-1]
      if s[i]==s[j]:
        j+=1
      next[i]=j
    return next[n-1]!=0 and n%(n-next[n-1])==0